kernel/examples/inspect-table/src/main.rs

```diff
@@ -77,97 +81,107 @@ impl Action {
     }
 }
 
-fn fields_in(field: &StructField) -> usize {
-    if let DataType::Struct(ref inner) = field.data_type {
-        inner.fields().map(fields_in).sum()
-    } else {
-        1
-    }
-}
-
 struct LogVisitor {
     actions: Vec<Action>,
     add_offset: usize,
     remove_offset: usize,
     protocol_offset: usize,
     metadata_offset: usize,
     set_transaction_offset: usize,
+    commit_info_offset: usize,
     previous_rows_seen: usize,
 }
 
 impl LogVisitor {
-    fn new(log_schema: &SchemaRef) -> LogVisitor {
-        let mut offset = 0;
-        let mut add_offset = 0;
-        let mut remove_offset = 0;
-        let mut protocol_offset = 0;
-        let mut metadata_offset = 0;
-        let mut set_transaction_offset = 0;
-        for field in log_schema.fields() {
-            match field.name().as_str() {
-                "add" => add_offset = offset,
-                "remove" => remove_offset = offset,
-                "protocol" => protocol_offset = offset,
-                "metaData" => metadata_offset = offset,
-                "txn" => set_transaction_offset = offset,
-                _ => {}
-            }
-            offset += fields_in(field);
-        }
+    fn new() -> LogVisitor {
+        // NOTE: Each `position` call consumes the first item of the searched-for batch. So we have
+        // to add one to each `prev_offset`. But the first call starts from actual first, so we
+        // manually skip the first entry to make the first call behavior match the other calls.
+        let mut names = NAMES_AND_TYPES.as_ref().0.iter();
+        names.next();
+
+        let mut next_offset =
+            |prev_offset, name| prev_offset + 1 + names.position(|n| n[0] == name).unwrap();
+        let add_offset = 0;
```

Comment:
This is a lot cleaner! One question I have is whether we can do this work using macros. Something like 
```
struct XVisitor {
    AddVisitor,
    RemoveVisitor
}
```
that would generate all this code.


The reason I ask is we keep implementing visitors for the same fields:
- [Visitor for adds or removes](https://github.com/delta-io/delta-kernel-rs/blob/main/kernel/src/scan/log_replay.rs#L46)
- [The ScanFile visitor for Scan](https://github.com/delta-io/delta-kernel-rs/blob/main/kernel/src/scan/state.rs#L145)
- [The ScanFile visitor for TableChanges](https://github.com/delta-io/delta-kernel-rs/pull/474/files#diff-c67c278b9eadf5309edc95b2a41e7823b989ef19a2001ea9f95ec9728d870be2R87)
- [Visitor for Add, remove, and Cdc](https://github.com/delta-io/delta-kernel-rs/pull/474/files#diff-03dadaf7616d2cefb3d9027051970d780aa395977cae25b01aa84415e1922017R347)

Reply:
Or perhaps some way to "project" the `LogVisitor`?

Reply:
> This is a lot cleaner! One question I have is whether we can do this work using macros... that would generate all this code?

It's a good question... for visitors that just deserialize the row to some concrete struct, a derive macro should be able to handle the boilerplate. But for visitors that are doing something else... seems arbitrarily complex/difficult to capture that in a macro?

There's also the problem of identifying whether a given sub-visitor should be visited. E.g. we use `add.path` because it's required in the `Add` schema, but what if we were to ever support visitors for `CommitInfo` where _all_ columns are optional?

> we keep implementing visitors for the same fields:
> 
>     * [Visitor for adds or removes](https://github.com/delta-io/delta-kernel-rs/blob/main/kernel/src/scan/log_replay.rs#L46)

This one goes away in https://github.com/delta-io/delta-kernel-rs/pull/494, and becomes something very custom (tho with only a few fields, so simpler than before).

>     * [The ScanFile visitor for Scan](https://github.com/delta-io/delta-kernel-rs/blob/main/kernel/src/scan/state.rs#L145)

This resembles the Add visitor, but the schema is not actually the same (even tho it's very similar). I don't know how easily a macro could deal with that? 

>     * [The ScanFile visitor for TableChanges](https://github.com/delta-io/delta-kernel-rs/pull/474/files#diff-c67c278b9eadf5309edc95b2a41e7823b989ef19a2001ea9f95ec9728d870be2R87)
> 
>     * [Visitor for Add, remove, and Cdc](https://github.com/delta-io/delta-kernel-rs/pull/474/files#diff-03dadaf7616d2cefb3d9027051970d780aa395977cae25b01aa84415e1922017R347)

This one is annoying, for the same reason this inspect-table example is annoying (or the Add+Remove visitor mentioned earlier): It's really just dispatching to other sub-visitors to do the real work.